server/middleware/RegistryUtils.js

```diff
@@ -61,34 +61,60 @@ const normalizeTarHeader = (header) => {
   return header
 }
 
-const getPackage = (tarballURL, outputDir, callback) => {
-  mkdirp(outputDir, (error) => {
+const extractions = {}
+
+const triggerCallbacks = (tarballURL, error) => {
+  const callbacks = extractions[tarballURL]
+
+  for (let i = 0; i < callbacks.length; i++) {
     if (error) {
-      callback(error)
+      callbacks[i](error)
     } else {
-      let callbackWasCalled = false
-
-      fetch(tarballURL).then(response => {
-        response.body
-          .pipe(gunzip())
-          .pipe(
-            tar.extract(outputDir, {
-              dmode: 0o666, // All dirs should be writable
-              fmode: 0o444, // All files should be readable
-              map: normalizeTarHeader
-            })
-          )
-          .on('finish', callback)
-          .on('error', (error) => {
-            if (callbackWasCalled) // LOL node streams
-              return
-
-            callbackWasCalled = true
-            callback(error)
-          })
-      })
+      callbacks[i]()
     }
-  })
+  }
+}
+
+const getPackage = (tarballURL, outputDir, callback) => {
+  if (!extractions[tarballURL]) {
+    extractions[tarballURL] = [];
+  }
+
+  if (!extractions[tarballURL].length) {
+    extractions[tarballURL].push(callback)
```

Comment:
Instead of keeping an array of callbacks I think we can just convert the whole fetch + extract operation into a promise and return that same promise object when we get a new request for it. This should make the code a bit more straightforward and eliminate the need to loop over and call all the callbacks when we're finished. I'll take a look at this today and run your gist to make sure it works.